File: hackerrank/cracking_the_coding_interview/data_structures/trees_binary_search/is_binary_search_tree.rb
class Node
  attr_accessor :left, :right
  attr_reader :data
  
  def initialize(data = nil)
    @data = data
    @left = nil
    @right = nil
  end 
end

def BFS(node, data_hash)
  return false if node.nil?
  
end

def checkBST(root):
  return BFS(root, -1, 1 * 10^4)
   

# Test case tree
a = Node.new(4)
b = Node.new(2)
c = Node.new(1)
d = Node.new(3)
e = Node.new(6)
f = Node.new(5)
g = Node.new(5)

a.left = b
a.right = e
b.left = c
b.right = d
e.left = f
e.right = g

File: hackerrank/cracking_the_coding_interview/data_structures/trees_binary_search/is_binary_search_tree.py
class node:
    def __init__(self, data):
        self.data = data
        self.left = None
        self.right = None

import sys

def BFS(node, min, max):
  # Return true if tree is empty
  if node is None:
    return True
  
  # Check against min and max constraints
  if node.data < min or node.data > max:
    return False
  
  #Else recursively check subtrees and update min and max
  return BFS(node.left, min, node.data - 1) and BFS(node.right, node.data + 1, max)

def checkBST(root):
  return BFS(root, sys.maxsize * -1, sys.maxsize)
   